addons/godot-next/nodes/callback_delegator.gd

```diff
@@ -0,0 +1,163 @@
+# author: xdgamestudios
+# license: MIT
+# description: Manages a ResourceSet of element resources and delegates Node callbacks to each instance.
+#              As a ResourceSet, only one element of any given type is allowed on a single elements node.
+# deps:
+# - ResourceSet
+# - PropertyInfo
+# - element
+# - ClassType
+# usage:
+# - Creating:
+#     elements = elements.new()
+# - Adding elements:
+#     elements.add_element(Myelement) # Returns a new or pre-existing instance of the element or null if given an invalid element script.
+# - Checking elements:
+#     elements.has_element(Myelement) # Returns true if the element exists in the collection.
+# - Retrieving elements:
+#     elements.get_element(Myelement) # Returns the element instance of the given type or null if not in the collection.
+# - Removing elements:
+#     elements.remove_element(Myelement) # Removes the element from the collection. Returns true if successful. Else, returns false.
+tool
+extends Node
+class_name CallbackDelegator
+
+##### PROPERTIES #####
+
+var _elements: ResourceSet = null
+
+var _callbacks: Dictionary = {
+	"_enter_tree" : {},
+	"_exit_tree" : {},
+	"_ready" : {},
+	"_process" : {},
+	"_physics_process" : {},
+	"_input" : {},
+	"_unhandled_input" : {},
+	"_unhandled_key_input" : {}
+}
+
+##### NOTIFICATIONS #####
+
+func _get_property_list() -> Array:
+	if not _elements:
+		_elements = ResourceSet.new()
+	return [ PropertyInfo.new_resource("_elements").to_dict() ]
+
+func _ready() -> void:
+	_handle_notification("_ready")
+
+func _enter_tree() -> void:
+	var elements = _elements.get_data().values()
+	for element in elements:
+		if not element.get_owner():
+			_initialize_element(element)
+	_check_for_empty_callbacks()
+	
+	_handle_notification("_enter_tree")
+
+func _exit_tree() -> void:
+	_handle_notification("_enter_tree")
+	pass
+
+func _process(delta: float) -> void:
+	_handle_notification("_process", delta)
+
+func _physics_process(delta: float) -> void:
+	_handle_notification("_physics_process", delta)
+
+func _input(event: InputEvent) -> void:
+	_handle_notification("_input", event)
+
+func _unhandled_input(event: InputEvent) -> void:
+	_handle_notification("_unhandled_input", event)
+
+func _unhandled_key_input(event: InputEventKey) -> void:
+	_handle_notification("_unhandled_key_input", event)
+
+##### VIRTUALS #####
+
+##### PUBLIC METHODS #####
+
+func add_element(p_type: Script) -> Resource:
+	var elements = _elements.get_data()
+	
+	var ct = ClassType.new(p_type)
```

Comment:
Since you also use ClassType here, you should put it in the CallbackDelegator and then assign the `ResourceSet`'s ClassType reference to be the same as the CallbackDelegator's.

    `elements.ct = ct`

Reply:
although, they should probably be private variables.

Reply:
Or, if you want some additional safety, just create a new instance and pass in the CallbackDelegator's instance to initiate the type's copy constructor

    elements._ct = ClassType.new(_ct) # creates a new ClassType that copies the reference to the internal script map.